### PR TITLE
Require AddExplicitInterfaceImplementation for adding a type that implements member explicitly

### PR DESCRIPTION
**Pull request by tmat**
Adding a new row to InterfaceImpl table may case AV on .NET Framework runtime. Report rude edit to prevent that from happening.
The rude edit is reported when `AddExplicitInterfaceImplementation` runtime capability is not present.

Contributes to fix of https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1452066.
Debugger part: https://devdiv.visualstudio.com/DevDiv/_git/Concord/pullrequest/546202

### CODE REVIEW

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueCapabilitiesGrantor.cs [hunk 1]
```diff
@@ -18,4 +18,8 @@ public bool Grant(EditAndContinueCapabilities capabilities)
         GrantedCapabilities |= capabilities;
         return (_availableCapabilities & capabilities) == capabilities;
     }
+
+    public bool GrantNewTypeDefinition(INamedTypeSymbol type)
+        => Grant(EditAndContinueCapabilities.NewTypeDefinition) &&
+           (!type.HasExplicitInterfaceImplementation() || Grant(EditAndContinueCapabilities.AddExplicitInterfaceImplementation));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EditAndContinue/EditAndContinueCapabilitiesGrantor.cs:24]: i coudln't understand this at all.  Consider breaking into statements :)

File: src/Features/Core/Portable/EditAndContinue/Utilities/Extensions.cs [hunk 2]
```diff
@@ -226,4 +226,10 @@ private static bool ParsePrimaryParameterBackingFieldName(string fieldName, [Not
 
     public static ISymbol PartialAsImplementation(this ISymbol symbol)
         => symbol is IMethodSymbol { PartialImplementationPart: { } impl } ? impl : symbol;
+
+    /// <summary>
+    /// Returns true if any member of the type implements an interface member explicitly.
+    /// </summary>
+    public static bool HasExplicitInterfaceImplementation(this INamedTypeSymbol type)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EditAndContinue/Utilities/Extensions.cs:233]: nit.  Consider: AnyMemberHasExplicitInterfaceImplementations

**Comment by tmat** [comment]
@davidwengier ptal

**Review by davidwengier — APPROVED** [review]
Given the number of tests that have the new capability, I'm surprised this hasn't come up in the past. Or maybe it has :)

**Comment by tmat** [comment]
@davidwengier The tests do not actually apply the change. The AV is from .NET Framework and it does not occur if the change is the first applied change.

**Comment by davidwengier** [comment]
Yeah, I just took the number of tests as an indication of the number of scenarios that could cause an AV. If the ordering is also important, then that could explain it for sure.

**Comment by tmat** [comment]
Waiting for more customer validation from 17.11 previews before merging.

**Comment by lambdageek** [comment]
Is there a corresponding runtime change or a GH issue to light this up for CoreCLR and Mono? /cc @mikelle-rogers

**Pull request merged**